File: 42-trapping-rain-water/trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int i;
    int n=height.size();
    vector<int>l(n);
    vector<int>r(n);
    int max=0;
    for(i=0;i<n;i++)
    {
        if(max<height[i])
        {
           max=height[i]; 
        }
        l[i]=max;
    }
    max=0;
    for(i=n-1;i>=0;i--)
    {
        if(max<height[i])
        {
            max=height[i];
        }
        r[i]=max;
    }
    int t=0;
    for(i=0;i<n;i++)
    {
        t=t+min(l[i],r[i])-height[i];
    }
    return t;
        
    }
};

File: 42-trapping-rain-water/trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int n=height.length;
        int i=0,j=(n-1);
        int l=height[i],r=height[j];
        int res=0;
        while(i<j)
        {
            if(l<r)
            {
               i=i+1;
               l=Math.max(l,height[i]);
               res+=l-height[i];
            }
            else
            {
                j=j-1;
                r=Math.max(r,height[j]);
                res+=r-height[j];
            }
        }
        return res;
        

        
    }
}